krun/util.py

```diff
@@ -186,10 +199,17 @@ def read_popen_output_carefully(process, platform, print_stderr=True, timeout=No
         stderr_consumer = None
 
     while open_fds:
-        ready = select.select(open_fds, [], [], SELECT_TIMEOUT)
-
         if timeout_at is not None:
-            debug("select done: %s" % datetime.now())
+            time_left = (timeout_at - datetime.now()).total_seconds()
+            if time_left <= 0:
+                raise ExecutionTimeout()
+            ready = select.select(open_fds, [], [], time_left)
+        else:
+            # The `select()` documentation doesn't specify what to pass for
+            # ulimited timeout. It only says that the parameter should be
```

Comment:
Surely we don't want unlimited (I assume "ulimited" is a typo!) timeout? We only want to wait until at most the end of the timeout period?

Reply:
If it's unlimited, then this PR doesn't do anything in the common case. At the moment we only apply the timeout after stdout/stderr have been closed, but generally they're closed only at the point a process is going to exit anyway, so that doesn't really buy us much. What you normally want to do is run the process for at most `timeout`, bearing in mind that `poll`/`select` can return in the interim so you need to keep updating the timeout. Something like this:

```python
timeout = <some point in the future>
while time.time() < timeout:
    remaining = min(timeout - time.time(), 0)
    poll(fds, remaining)

poll.wait(min(timeout - time.time(), 0))
```

For example, [here's lang_tester doing this sort of thing](https://github.com/softdevteam/lang_tester/blob/master/src/tester.rs#L710) (I admit that it's a bit hard to read because it's using the C API).

Reply:
> If it's unlimited, then this PR doesn't do anything in the common case.

This function defaults to no timeout, if that's what you mean? We use this function for all kinds of other things too, not just for process executions, so we don't want to affect those.

> At the moment we only apply the timeout after stdout/stderr have been closed

Eh? I'm not sure how you've arrived at that. 

```
> /home/edd/research/krun/krun/util.py(207)read_popen_output_carefully()
-> ready = select.select(open_fds, [], [], time_left)
(Pdb) open_fds
[4, 6]
(Pdb) timeout
30
(Pdb) time_left
30.0
```

stderr and stdout are both open and select is being called with a 30 second timeout.

Reply:
> This function defaults to no timeout, if that's what you mean?

Isn't the point of this PR to enforce a timeout? I feel like I'm missing out on something!

Reply:
Does this clarify?
https://github.com/softdevteam/krun/pull/396#discussion_r423865174

Reply:
Yeah, because some other unrelated bits of code (for which it wouldn't make sense to have a timeout) use this routine too.

So if this all seems reasonable, I'll hook this into the config file and results file. I just wanted to check the overall approach looks sane!